svn/trunk/SlickEngagement_Plugin.php

```diff
@@ -293,135 +262,199 @@ public function removeSemicolons($value)
+
+        // Check for existing data in transient cache. If none, then fetch data from server.
+        if (false === ($boot_data_obj = get_transient($transient_name))) {
+            $this->echoSlickstreamComment("Fetching page boot data from server");
+            $boot_data_obj = $this->fetchBootData($siteCode);
+
+            //TODO: store cache hits and cache misses in a transient or option?
+
+	        // Put the results in transient storage; expire after 15 minutes
+            if ($boot_data_obj) {
+                set_transient($transient_name, $boot_data_obj, 15 * MINUTE_IN_SECONDS);
+            } else {
+                $this->echoSlickstreamComment("Error Fetching page boot data from server");
+                return;
             }
-          }
+        } else {
+            $this->echoSlickstreamComment("Using cached page boot data");
         }
-      }
+
+        $this->echoSlickBootJs($boot_data_obj);
+        $this->echoClsData($boot_data_obj);
     }
 
+    //Plugin-based A/B Testing JS Logic
+    private function getAbTestJs()
+    {
+        
+        $jsBlock = <<<JSBLOCK
+        window.slickAbTestResult = function(percentEnabled, recalculate = false, testName = 'embed') {
+        const win = window;
+        const storage = win.localStorage;
+        const targetPercentEnabled = parseInt(percentEnabled);
+        
+        if (isNaN(targetPercentEnabled)) {
+            return new Error("Invalid enabled percentage");
+        }
+
+        let enableSlickFeature;
+        const abTestStorageKey = `slickab-\${testName}-\${targetPercentEnabled}`;
+        const storedOnOffVal = storage.getItem(abTestStorageKey);
+        
+        const percentKey = `slickAbTestPercent-\${testName}`;
+        const storedPercentVal = parseInt(storage.getItem(percentKey));
+        
+        if (recalculate === true || !storedOnOffVal || storedPercentVal !== targetPercentEnabled) {
+            enableSlickFeature = (Math.random() * 100) <= targetPercentEnabled;
+            storage.setItem(abTestStorageKey, enableSlickFeature);
+            storage.setItem(percentKey, targetPercentEnabled);
+        } else {
+            enableSlickFeature = storage.getItem(abTestStorageKey) === 'true';
+        }
+
+        const abGroupVal = `slk\${testName}\${targetPercentEnabled}`;
+        const featureOnOff = enableSlickFeature ? "on" : "off";
+        win.adthrive = win.adthrive || {};
+        win.adthrive.cmd = win.adthrive.cmd || [];
+        win.adthrive.cmd.push(() => { win.adthrive.config.abGroup.set(abGroupVal, featureOnOff); });
+
+        return enableSlickFeature;
+        };
+        JSBLOCK;
+        
+        return $jsBlock;
+    }
+
+    //TODO: Clean this up / migrate to SR functions
     public function addSlickPageHeader()
     {
         global $post;
 
-        $this->getPageBootData();
+        $this->echoSlickstreamComment("Page Generated at " . $this->getCurrentTimeStampByTimeZone('America/New York') . " EST");
```

Comment:
I tried running this plugin change, and got this error.  I would suggest testing it before it is ready for code review.

Fatal error: Uncaught Exception: DateTimeZone::__construct(): Unknown or bad timezone (America/New York)

Reply:
I bet it should be "America/New_York"